Added support for nodes dialing back to cluster.

russjones opened the pull request:
**Description**

Updated `services.ReverseTunnel` to support type (proxy or node). For proxy types, which represent trusted cluster connections, when a `services.ReverseTunnel` is created, it's created on the remote side with name `/reverseTunnels/example.com`. For node types, `services.ReverseTunnel` is created on the main side as `/reverseTunnels/{nodeUUID}.clusterName`.

Updated `services.TunnelConn` to support type (proxy or node). For proxy types, which represent trusted cluster connections, tunnel connections are created on the main side under `/tunnelConnections/remote.example.com/{proxyUUID}-remote.example.com`. For nodes, tunnel connections are created on the main side under `/tunnelConnections/example.com/{proxyUUID}-example.com`. This allows searching for tunnel connections by cluster then allows easily creating a set of proxies that are missing matching `services.TunnelConn`.

The reverse tunnel server has been updated to handle heartbeats from proxies as well as nodes. Proxy heartbeat behavior has not changed. Heartbeats from nodes now add remote connections to the matching local site. In addition, the reverse tunnel server now proxies connection to the Auth Server for requests that are already authenticated (a second authentication to the Auth Server is required).

For registration, nodes try and connect to the Auth Server to fetch host credentials. Upon failure, nodes now try and fallback to fetching host credentials from the web proxy.

To establish a connection to an Auth Server, nodes first try and connect directly, and if the connection fails, fallback to obtaining a connection to the Auth Server through the reverse tunnel. If a connection is established directly, node startup behavior has not changed. If a node establishes a connection through the reverse tunnel, it creates an AgentPool that attempts to dial back to the cluster and establish a reverse tunnel.

When nodes heartbeat, they also heartbeat if they are connected directly to the cluster or through a reverse tunnel. For nodes that are connected through a reverse tunnel, the proxy subsystem now directs the reverse tunnel server to establish a connection through the reverse tunnel instead of directly.

When sending discovery requests, the domain field has been replaced with tunnelID. The tunnelID field is either the cluster name (same as before) for proxies, or `{nodeUUID}.example.com` for nodes.

**Related Issues**

Fixes https://github.com/gravitational/teleport/issues/803

klizhentas commented:
>  For node types, services.ReverseTunnel is created on the main side as /reverseTunnels/{nodeUUID}.clusterName

I'm pretty sure this not going to be scalable for 1K+ clusters

russjones commented on lib/service/connect.go:
```diff
@@ -742,7 +778,125 @@ func (process *TeleportProcess) rotate(conn *Connector, localState auth.StateV2,
+
+	// Connect to the Auth Server through the reverse tunnel server.
+	authDialer := func(in context.Context, network, addr string) (net.Conn, error) {
+		channel, _, err := conn.OpenChannel(reversetunnel.ChanTransport, nil)
```
`reversetunnel.Dial`

russjones commented on lib/reversetunnel/srv.go:
```diff
@@ -490,26 +493,95 @@ func (s *server) HandleNewChan(conn net.Conn, sconn *ssh.ServerConn, nch ssh.New
+
+	// Extract the role. For proxies, it's another cluster asking to join, for
+	// nodes it's a node dialing back.
+	val, ok := sconn.Permissions.Extensions["role"]
```
constant

russjones commented on lib/reversetunnel/localsite.go:
```diff
@@ -216,18 +240,149 @@ func (s *localSite) dialWithAgent(params DialParams) (net.Conn, error) {
+		},
+	)
+	tunnelConn.SetLastHeartbeat(t)
+	tunnelConn.SetExpiry(time.Now().Add(defaults.ReverseTunnelOfflineThreshold))
```
Use same delay as `services.ReverseTunnel`.

russjones commented on lib/reversetunnel/localsite.go:
```diff
@@ -216,18 +240,149 @@ func (s *localSite) dialWithAgent(params DialParams) (net.Conn, error) {
 	}
-	return nil, trace.NotFound("server %v is unknown", addr)
+
+	rconn, ok := s.remoteConns[addr]
```
https://github.com/gravitational/teleport/blob/master/lib/reversetunnel/remotesite.go#L178

russjones commented on lib/reversetunnel/conn.go:
```diff
@@ -0,0 +1,500 @@
+	// If the request is for the remote auth server or Kubernetes proxy, resolve
+	// and connect to them. Otherwise connect to the passed in server.
+	switch server {
+	case RemoteAuthServer:
```
Comment.

russjones commented on lib/reversetunnel/conn.go:
```diff
@@ -0,0 +1,500 @@
+	req.Reply(true, []byte("Connected."))
+	p.log.Debugf("Successfully dialed to %v, start proxying.", server)
+
+	wg := sync.WaitGroup{}
```
Context.

russjones commented on lib/reversetunnel/discovery.go:
```diff
@@ -11,10 +11,21 @@ import (
+	// TunnelID identifies who the tunnel is connected to. For trusted clusters,
+	// the TunnelID is the name of the remote cluster (like example.com). For
+	// nodes, it is the nodeID (like 4a050852-23b5-4d6d-a45f-bed02792d453.example.com).
+	TunnelID string `json:"cluster_name"`
```
Backward compat.

russjones commented on lib/srv/regular/proxy.go:
```diff
@@ -287,8 +287,15 @@ func (t *proxySubsys) proxyToHost(
+		// Check if server requested to be connected to over a reverse tunnel.
+		if server.GetUseTunnel() {
+			useTunnel = true
+			serverAddr = fmt.Sprintf("%v.%v", server.GetName(), t.clusterName)
```
See if this can be removed and figured out in the site.

russjones commented on lib/service/service.go:
```diff
@@ -1421,12 +1472,48 @@ func (process *TeleportProcess) initSSH() error {
+		select {
+		case <-process.ExitContext().Done():
+			return
+		case <-time.Tick(defaults.UpsertTunnelHeartbeatTick):
```
Match rate.

speeltronics commented:
@russjones  I compiled the branch rjones/iot5 on both my client and my server and I cannot get this to work. The way I was doing this by creating a cluster for each client is no longer working either. 
I can see the node on my server, but when I click to connect to it I just get:

disconnected
failed connecting to node MYNODE.
ERROR REPORT:
Original Error: *trace.ConnectionProblemError MYNODE is offline
Stack Trace:
        ~/go/src/github.com/gravitational/teleport/lib/reversetunnel/remotesite.go:556 github.com/gravitational/teleport/lib/reversetunnel.(*remoteSite).connThroughTunnel
        ~/go/src/github.com/gravitational/teleport/lib/reversetunnel/remotesite.go:482 github.com/gravitational/teleport/lib/reversetunnel.(*remoteSite).DialTCP
        ~/go/src/github.com/gravitational/teleport/lib/reversetunnel/remotesite.go:476 github.com/gravitational/teleport/lib/reversetunnel.(*remoteSite).Dial
        ~/go/src/github.com/gravitational/teleport/lib/srv/regular/proxy.go:320 github.com/gravitational/teleport/lib/srv/regular.(*proxySubsys).proxyToHost
        ~/go/src/github.com/gravitational/teleport/lib/srv/regular/proxy.go:190 github.com/gravitational/teleport/lib/srv/regular.(*proxySubsys).Start
        ~/go/src/github.com/gravitational/teleport/lib/srv/regular/sshserver.go:1145 github.com/gravitational/teleport/lib/srv/regular.(*Server).handleSubsystem
        ~/go/src/github.com/gravitational/teleport/lib/srv/regular/sshserver.go:1022 github.com/gravitational/teleport/lib/srv/regular.(*Server).dispatch
        ~/go/src/github.com/gravitational/teleport/lib/srv/regular/sshserver.go:987 github.com/gravitational/teleport/lib/srv/regular.(*Server).handleSessionRequests
        /usr/local/go/src/runtime/asm_amd64.s:2362 runtime.goexit
User Message: MYNODE is offline

Do you have example yaml files for the server and client or a cli I can use to test this? Thanks!

speeltronics commented:
Could it be the roles to use the username on the client isn't working on this branch...

I'm getting this on the client.

Access to read role in namespace default denied to roles RemoteProxy,default-implicit-role: no allow rule matched. services/role.go:1720

speeltronics commented:
I still cannot get this to work. If I connect a node to my my main node running the auth server it will show up on the web panel, but when I try to connect to it, I get connected to the main node running the auth server. The same thing happens whether my main auth server is runnning locally or remotely. I also tried specifying the actual node ip on the ssh instead of using 0.0.0.0, but then it timesout and never connects.

russjones committed:
Added support for nodes dialing back to cluster.

Updated services.ReverseTunnel to support type (proxy or node). For
proxy types, which represent trusted cluster connections, when a
services.ReverseTunnel is created, it's created on the remote side with
name /reverseTunnels/example.com. For node types, services.ReverseTunnel
is created on the main side as /reverseTunnels/{nodeUUID}.clusterName.

Updated services.TunnelConn to support type (proxy or node). For proxy
types, which represent trusted cluster connections, tunnel connections
are created on the main side under
/tunnelConnections/remote.example.com/{proxyUUID}-remote.example.com.
For nodes, tunnel connections are created on the main side under
/tunnelConnections/example.com/{proxyUUID}-example.com. This allows
searching for tunnel connections by cluster then allows easily creating
a set of proxies that are missing matching services.TunnelConn.

The reverse tunnel server has been updated to handle heartbeats from
proxies as well as nodes. Proxy heartbeat behavior has not changed.
Heartbeats from nodes now add remote connections to the matching local
site. In addition, the reverse tunnel server now proxies connection to
the Auth Server for requests that are already authenticated (a second
authentication to the Auth Server is required).

For registration, nodes try and connect to the Auth Server to fetch host
credentials. Upon failure, nodes now try and fallback to fetching host
credentials from the web proxy.

To establish a connection to an Auth Server, nodes first try and connect
directly, and if the connection fails, fallback to obtaining a
connection to the Auth Server through the reverse tunnel. If a
connection is established directly, node startup behavior has not
changed. If a node establishes a connection through the reverse tunnel,
it creates an AgentPool that attempts to dial back to the cluster and
establish a reverse tunnel.

When nodes heartbeat, they also heartbeat if they are connected directly
to the cluster or through a reverse tunnel. For nodes that are connected
through a reverse tunnel, the proxy subsystem now directs the reverse
tunnel server to establish a connection through the reverse tunnel
instead of directly.

When sending discovery requests, the domain field has been replaced with
tunnelID. The tunnelID field is either the cluster name (same as before)
for proxies, or {nodeUUID}.example.com for nodes.

russjones commented:
@gravitational-jenkins retest this please
